pkg/assembler/assembler.go

```diff
@@ -12,145 +12,104 @@ var parser *participle.Parser[CasmProgram] = participle.MustBuild[CasmProgram](
+	// Ast To Instruction List
+	instructionList, err := astToInstruction(casmAst)
+	if err != nil {
+		return nil, err
+	}
+	// Instruction to bytecode
+	return encodeInstructionListToBytecode(instructionList)
 }
 
-//
-// Functions that visit the AST in order to encode the instructions
-//
-
-const (
-	// offsets
-	op0Offset = 16
-	op1Offset = 32
-
-	// flag values
-	dstRegBit         = 48
-	op0RegBit         = 49
-	op1ImmBit         = 50
-	op1FpBit          = 51
-	op1ApBit          = 52
-	resAddBit         = 53
-	resMulBit         = 54
-	pcJumpAbsBit      = 55
-	pcJumpRelBit      = 56
-	pcJnzBit          = 57
-	apAddBit          = 58
-	apAdd1Bit         = 59
-	opcodeCallBit     = 60
-	opcodeRetBit      = 61
-	opcodeAssertEqBit = 62
-
-	// default values
-	biasedZero     uint16 = 0x8000
-	biasedPlusOne  uint16 = 0x8001
-	biasedMinusOne uint16 = 0x7FFF
-	biasedMinusTwo uint16 = 0x7FFE
-)
-
-func encodeCasmProgram(casmAst CasmProgram) ([]*f.Element, error) {
-	n := len(casmAst.Instructions)
-	bytecode := make([]*f.Element, 0, n+(n/2)+1)
-	var err error
-	for i := range casmAst.Instructions {
-		bytecode, err = encodeInstruction(bytecode, casmAst.Instructions[i])
+/*
+*    Casm to instruction list functions
+ */
+func astToInstruction(ast *CasmProgram) ([]Instruction, error) {
+	// Vist ast
+	n := len(ast.Ast)
+	// Slice with length 0 and capacity n
+	instructionList := make([]Instruction, 0, n)
+	// iterate over the AST
+	for i := range ast.Ast {
+		instruction, err := encodeNodeToInstr(ast.Ast[i])
 		if err != nil {
 			return nil, err
 		}
+		// Append instruction to list
+		instructionList = append(instructionList, instruction)
 	}
-	return bytecode, nil
+	return instructionList, nil
 }
 
-func encodeInstruction(bytecode []*f.Element, instruction Instruction) ([]*f.Element, error) {
-	var encode uint64 = 0
-	expression := instruction.Expression()
-
-	encode, err := encodeDstReg(&instruction, encode)
-	if err != nil {
-		return nil, err
-	}
-
-	encode, err = encodeOp0Reg(&instruction, expression, encode)
-	if err != nil {
-		return nil, err
-	}
-
-	encode, imm, err := encodeOp1Source(&instruction, expression, encode)
-	if err != nil {
-		return nil, err
-	}
-
-	encode = encodeResLogic(expression, encode) |
-		encodePcUpdate(instruction, encode) |
-		encodeApUpdate(instruction, encode) |
-		encodeOpCode(instruction, encode)
-
-	encodeAsFelt := new(f.Element).SetUint64(encode)
-
-	bytecode = append(bytecode, encodeAsFelt)
-	if imm != nil {
-		bytecode = append(bytecode, imm)
-	}
-
-	return bytecode, nil
+func encodeNodeToInstr(node AstNode) (Instruction, error) {
+	var instr Instruction
+	expr := node.Expression()
+	encodeDst(&node, &instr)
+	encodeOp0(&node, &instr, expr)
+	encodeOp1(&node, &instr, expr)
+	encodeFlags(&node, &instr, expr)
+	return instr, nil
 }
 
-func encodeDstReg(instr *Instruction, encode uint64) (uint64, error) {
-	if instr.ApPlus != nil || instr.Jump != nil {
+func encodeDst(node *AstNode, instr *Instruction) {
```

Comment:
I am not fond of prefixing the names with `encode...` since there is no encoding happening in this methods. I would suggest `setInstructionDst` or smth along the lines. This comment applies for the other functions with a similar purpose to this one. Tell me what you think

Reply:
yeah I agree. I've updated those functions